File: games/breakpoint/src/core/FixedStepDriver.ts
import { FIXED_DT } from '../physics/PhysicsConstants';
import type { PhysicsWorld } from '../physics/PhysicsWorld';

/**
 * The only bridge from wall-clock frames to simulation steps.
 *
 * Frames arrive at whatever rate the display and the machine manage. The
 * simulation must not care. This accumulates real elapsed time and spends it in
 * whole FIXED_DT steps, so a 30 fps laptop and a 144 fps desktop run the *same*
 * number of steps over the same simulated interval and reach the same result.
 * The leftover is exposed as `alpha` purely so the renderer can interpolate;
 * nothing physical ever reads it.
 *
 * `maxFrameSeconds` caps how much time one frame may spend. Without it, a tab
 * returning from the background hands over several seconds at once and the
 * catch-up loop stalls the page — the classic spiral of death.
 */
export class FixedStepDriver {
  private accumulator = 0;
  /** Fraction of a step left over, in [0, 1). Render interpolation only. */
  alpha = 0;
  /** Steps executed on the most recent `advance()`. */
  lastStepCount = 0;

  constructor(
    private readonly world: PhysicsWorld,
    private readonly maxFrameSeconds = 0.25,
  ) {}

  /** Feed one frame's elapsed seconds. Returns the number of steps run. */
  advance(frameSeconds: number): number {
    if (!Number.isFinite(frameSeconds) || frameSeconds <= 0) {
      this.lastStepCount = 0;
      return 0;
    }
    this.accumulator += Math.min(frameSeconds, this.maxFrameSeconds);

    let steps = 0;
    while (this.accumulator >= FIXED_DT) {
      this.world.step();
      this.accumulator -= FIXED_DT;
      steps++;
    }
    this.alpha = this.accumulator / FIXED_DT;
    this.lastStepCount = steps;
    return steps;
  }

  reset(): void {
    this.accumulator = 0;
    this.alpha = 0;
    this.lastStepCount = 0;
  }
}

File: games/breakpoint-unity/Assets/BREAKPOINT/Runtime/Simulation/FixedStepDriver.cs
using System;

namespace Breakpoint.Simulation
{
    /// <summary>
    /// The only bridge from wall-clock frames to simulation steps.
    ///
    /// Frames arrive at whatever rate the display and the device manage. The
    /// simulation must not care. This accumulates real elapsed time and spends
    /// it in whole <see cref="PhysicsConstants.FixedDt"/> steps, so a 30 fps
    /// phone and a 120 fps tablet run the same number of steps over the same
    /// simulated interval and reach the same result. The leftover is exposed as
    /// <see cref="Alpha"/> purely so presentation can interpolate; nothing
    /// physical ever reads it.
    ///
    /// <see cref="MaxFrameSeconds"/> caps how much time one frame may spend.
    /// Without it, an app returning from the background hands over several
    /// seconds at once and the catch-up loop stalls the game — the classic
    /// spiral of death.
    ///
    /// This lives in the simulation assembly, not in a MonoBehaviour, so a
    /// headless test can drive frames without loading a scene.
    /// </summary>
    public sealed class FixedStepDriver
    {
        private readonly PhysicsWorld _world;
        private readonly double _maxFrameSeconds;
        private double _accumulator;

        public FixedStepDriver(PhysicsWorld world, double maxFrameSeconds = 0.25)
        {
            _world = world ?? throw new ArgumentNullException(nameof(world));
            _maxFrameSeconds = maxFrameSeconds;
        }

        /// <summary>Fraction of a step left over, in [0, 1). Presentation interpolation only.</summary>
        public double Alpha { get; private set; }

        /// <summary>Steps executed on the most recent <see cref="Advance"/>.</summary>
        public int LastStepCount { get; private set; }

        /// <summary>Feed one frame's elapsed seconds. Returns the number of steps run.</summary>
        public int Advance(double frameSeconds)
        {
            if (!Numeric.IsFinite(frameSeconds) || frameSeconds <= 0.0)
            {
                LastStepCount = 0;
                return 0;
            }

            _accumulator += Math.Min(frameSeconds, _maxFrameSeconds);

            int steps = 0;
            while (_accumulator >= PhysicsConstants.FixedDt)
            {
                _world.Step();
                _accumulator -= PhysicsConstants.FixedDt;
                steps++;
            }

            Alpha = _accumulator / PhysicsConstants.FixedDt;
            LastStepCount = steps;
            return steps;
        }

        public void Reset()
        {
            _accumulator = 0.0;
            Alpha = 0.0;
            LastStepCount = 0;
        }
    }
}
